// x-pack/plugins/infra/public/store/remote/source/selectors.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const reselect_1 = require("reselect");
const remote_graphql_state_1 = require("../../../utils/remote_state/remote_graphql_state");
const sourceStatusGraphqlStateSelectors = remote_graphql_state_1.createGraphqlStateSelectors();
exports.selectSource = sourceStatusGraphqlStateSelectors.selectData;
exports.selectSourceConfiguration = reselect_1.createSelector(exports.selectSource, source => (source ? source.configuration : null));
exports.selectSourceLogAlias = reselect_1.createSelector(exports.selectSourceConfiguration, configuration => (configuration ? configuration.logAlias : null));
exports.selectSourceMetricAlias = reselect_1.createSelector(exports.selectSourceConfiguration, configuration => (configuration ? configuration.metricAlias : null));
exports.selectSourceFields = reselect_1.createSelector(exports.selectSourceConfiguration, configuration => (configuration ? configuration.fields : null));
exports.selectSourceStatus = reselect_1.createSelector(exports.selectSource, source => (source ? source.status : null));
exports.selectSourceLogIndicesExist = reselect_1.createSelector(exports.selectSourceStatus, sourceStatus => (sourceStatus ? sourceStatus.logIndicesExist : null));
exports.selectSourceMetricIndicesExist = reselect_1.createSelector(exports.selectSourceStatus, sourceStatus => (sourceStatus ? sourceStatus.metricIndicesExist : null));
exports.selectSourceIndexFields = reselect_1.createSelector(exports.selectSourceStatus, sourceStatus => (sourceStatus ? sourceStatus.indexFields : []));
exports.selectDerivedIndexPattern = reselect_1.createSelector(exports.selectSourceIndexFields, exports.selectSourceLogAlias, exports.selectSourceMetricAlias, (indexFields, logAlias, metricAlias) => ({
    fields: indexFields,
    title: `${logAlias},${metricAlias}`,
}));
