// x-pack/plugins/apm/common/constants.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SERVICE_NAME = 'context.service.name';
exports.SERVICE_AGENT_NAME = 'context.service.agent.name';
exports.SERVICE_LANGUAGE_NAME = 'context.service.language.name';
exports.REQUEST_URL_FULL = 'context.request.url.full';
exports.USER_ID = 'context.user.id';
exports.PROCESSOR_NAME = 'processor.name';
exports.PROCESSOR_EVENT = 'processor.event';
exports.TRANSACTION_DURATION = 'transaction.duration.us';
exports.TRANSACTION_TYPE = 'transaction.type';
exports.TRANSACTION_RESULT = 'transaction.result';
exports.TRANSACTION_NAME = 'transaction.name';
exports.TRANSACTION_ID = 'transaction.id';
exports.TRANSACTION_SAMPLED = 'transaction.sampled';
exports.TRACE_ID = 'trace.id';
exports.SPAN_START = 'span.start.us';
exports.SPAN_DURATION = 'span.duration.us';
exports.SPAN_TYPE = 'span.type';
exports.SPAN_NAME = 'span.name';
exports.SPAN_ID = 'span.id';
exports.SPAN_SQL = 'context.db.statement';
exports.SPAN_HEX_ID = 'span.hex_id';
// Parent ID for a transaction or span
exports.PARENT_ID = 'parent.id';
exports.ERROR_GROUP_ID = 'error.grouping_key';
exports.ERROR_CULPRIT = 'error.culprit';
exports.ERROR_LOG_MESSAGE = 'error.log.message';
exports.ERROR_EXC_MESSAGE = 'error.exception.message';
exports.ERROR_EXC_HANDLED = 'error.exception.handled';
