// x-pack/plugins/infra/public/pages/link_to/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var link_to_1 = require("./link_to");
exports.LinkToPage = link_to_1.LinkToPage;
var redirect_to_container_logs_1 = require("./redirect_to_container_logs");
exports.getContainerLogsUrl = redirect_to_container_logs_1.getContainerLogsUrl;
exports.RedirectToContainerLogs = redirect_to_container_logs_1.RedirectToContainerLogs;
var redirect_to_host_logs_1 = require("./redirect_to_host_logs");
exports.getHostLogsUrl = redirect_to_host_logs_1.getHostLogsUrl;
exports.RedirectToHostLogs = redirect_to_host_logs_1.RedirectToHostLogs;
var redirect_to_pod_logs_1 = require("./redirect_to_pod_logs");
exports.getPodLogsUrl = redirect_to_pod_logs_1.getPodLogsUrl;
exports.RedirectToPodLogs = redirect_to_pod_logs_1.RedirectToPodLogs;
var redirect_to_host_detail_1 = require("./redirect_to_host_detail");
exports.getHostDetailUrl = redirect_to_host_detail_1.getHostDetailUrl;
exports.RedirectToHostDetail = redirect_to_host_detail_1.RedirectToHostDetail;
var redirect_to_container_detail_1 = require("./redirect_to_container_detail");
exports.getContainerDetailUrl = redirect_to_container_detail_1.getContainerDetailUrl;
exports.RedirectToContainerDetail = redirect_to_container_detail_1.RedirectToContainerDetail;
var redirect_to_pod_detail_1 = require("./redirect_to_pod_detail");
exports.getPodDetailUrl = redirect_to_pod_detail_1.getPodDetailUrl;
exports.RedirectToPodDetail = redirect_to_pod_detail_1.RedirectToPodDetail;
