// x-pack/plugins/beats_management/public/components/table/assignment_schema.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const table_1 = require("./table");
exports.beatsListAssignmentOptions = [
    {
        grow: false,
        name: 'Unenroll selected',
        showWarning: true,
        warningHeading: 'Unenroll selected beats?',
        warningMessage: 'The selected Beats will no longer use central management',
        action: table_1.AssignmentActionType.Delete,
        danger: true,
    },
    {
        name: 'Set tags',
        grow: false,
        lazyLoad: true,
        panel: {
            id: 1,
            name: 'Assign tags',
        },
    },
];
exports.tagListAssignmentOptions = [
    {
        danger: true,
        grow: false,
        name: 'Remove tag(s)',
        showWarning: true,
        warningHeading: 'Remove tag(s)',
        warningMessage: 'Remove the tag?',
        action: table_1.AssignmentActionType.Delete,
    },
];
exports.tagConfigAssignmentOptions = [
    {
        danger: true,
        grow: false,
        name: 'Remove tag(s)',
        showWarning: true,
        warningHeading: 'Remove tag(s)',
        warningMessage: 'Remove the tag from the selected beat(s)?',
        action: table_1.AssignmentActionType.Delete,
    },
];
