// x-pack/plugins/beats_management/public/components/table/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var table_1 = require("./table");
exports.AssignmentActionType = table_1.AssignmentActionType;
exports.Table = table_1.Table;
var assignment_schema_1 = require("./assignment_schema");
exports.beatsListAssignmentOptions = assignment_schema_1.beatsListAssignmentOptions;
exports.tagConfigAssignmentOptions = assignment_schema_1.tagConfigAssignmentOptions;
var controls_1 = require("./controls");
exports.ControlBar = controls_1.ControlBar;
var table_type_configs_1 = require("./table_type_configs");
exports.BeatDetailTagsTable = table_type_configs_1.BeatDetailTagsTable;
exports.BeatsTableType = table_type_configs_1.BeatsTableType;
exports.TagsTableType = table_type_configs_1.TagsTableType;
