// x-pack/plugins/beats_management/public/components/table/table.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importDefault(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const constants_1 = require("../../../common/constants");
const controls_1 = require("./controls");
var AssignmentActionType;
(function (AssignmentActionType) {
    AssignmentActionType[AssignmentActionType["Add"] = 0] = "Add";
    AssignmentActionType[AssignmentActionType["Assign"] = 1] = "Assign";
    AssignmentActionType[AssignmentActionType["Delete"] = 2] = "Delete";
    AssignmentActionType[AssignmentActionType["Edit"] = 3] = "Edit";
    AssignmentActionType[AssignmentActionType["Reload"] = 4] = "Reload";
    AssignmentActionType[AssignmentActionType["Search"] = 5] = "Search";
})(AssignmentActionType = exports.AssignmentActionType || (exports.AssignmentActionType = {}));
const TableContainer = styled_components_1.default.div `
  padding: 16px;
`;
class Table extends react_1.default.Component {
    constructor(props) {
        super(props);
        this.resetSelection = () => {
            this.setSelection([]);
        };
        this.setSelection = (selection) => {
            this.setState({
                selection,
            });
        };
        this.state = {
            selection: [],
        };
    }
    render() {
        const { assignmentOptions, hideTableControls, items, kueryBarProps, type } = this.props;
        const pagination = {
            initialPageSize: constants_1.TABLE_CONFIG.INITIAL_ROW_SIZE,
            pageSizeOptions: constants_1.TABLE_CONFIG.PAGE_SIZE_OPTIONS,
        };
        const selectionOptions = hideTableControls
            ? null
            : {
                onSelectionChange: this.setSelection,
                selectable: () => true,
                selectableMessage: () => 'Select this beat',
                selection: this.state.selection,
            };
        return (react_1.default.createElement(TableContainer, null,
            !hideTableControls &&
                assignmentOptions && (react_1.default.createElement(controls_1.ControlBar, { itemType: type.itemType, assignmentOptions: assignmentOptions, kueryBarProps: kueryBarProps, selectionCount: this.state.selection.length })),
            react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
            react_1.default.createElement(eui_1.EuiInMemoryTable, { columns: type.columnDefinitions, items: items, itemId: "id", isSelectable: true, pagination: pagination, selection: selectionOptions, sorting: true })));
    }
}
exports.Table = Table;
