// x-pack/plugins/beats_management/public/components/table/table_type_configs.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const lodash_1 = require("lodash");
const moment_1 = tslib_1.__importDefault(require("moment"));
const react_1 = tslib_1.__importDefault(require("react"));
const connected_link_1 = require("../connected_link");
const tag_1 = require("../tag");
exports.BeatsTableType = {
    itemType: 'Beats',
    columnDefinitions: [
        {
            field: 'name',
            name: 'Beat name',
            render: (name, beat) => (react_1.default.createElement(connected_link_1.ConnectedLink, { path: `/beat/${beat.id}` }, name)),
            sortable: true,
        },
        {
            field: 'type',
            name: 'Type',
            sortable: true,
        },
        {
            field: 'full_tags',
            name: 'Tags',
            render: (value, beat) => (react_1.default.createElement(eui_1.EuiFlexGroup, { wrap: true, responsive: true, gutterSize: "xs" }, (lodash_1.sortBy(beat.full_tags, 'id') || []).map(tag => (react_1.default.createElement(eui_1.EuiFlexItem, { key: tag.id, grow: false },
                react_1.default.createElement(connected_link_1.ConnectedLink, { path: `/tag/edit/${tag.id}` },
                    react_1.default.createElement(tag_1.TagBadge, { tag: tag }))))))),
            sortable: false,
        },
        {
            // TODO: update to use actual metadata field
            field: 'config_status',
            name: 'Config Status',
            render: (value, beat) => {
                let color = 'success';
                let statusText = 'OK';
                let tooltipText = 'Beat successfully applied latest config';
                switch (beat.config_status) {
                    case 'UNKNOWN':
                        color = 'subdued';
                        statusText = 'Offline';
                        if (moment_1.default().diff(beat.last_checkin, 'minutes') >= 10) {
                            tooltipText = 'This Beat has not connected to kibana in over 10min';
                        }
                        else {
                            tooltipText = 'This Beat has not yet been started.';
                        }
                        break;
                    case 'ERROR':
                        color = 'danger';
                        statusText = 'Error';
                        tooltipText = 'Please check the logs of this Beat for error details';
                        break;
                }
                return (react_1.default.createElement(eui_1.EuiFlexGroup, { wrap: true, responsive: true, gutterSize: "xs" },
                    react_1.default.createElement(eui_1.EuiToolTip, { content: tooltipText },
                        react_1.default.createElement(eui_1.EuiHealth, { color: color }, statusText))));
            },
            sortable: false,
        },
        {
            field: 'full_tags',
            name: 'Last config update',
            render: (tags) => tags.length ? (react_1.default.createElement("span", null, moment_1.default(lodash_1.first(lodash_1.sortByOrder(tags, ['last_updated'], ['desc'])).last_updated).fromNow())) : null,
            sortable: true,
        },
    ],
    controlDefinitions: (data) => ({
        actions: [
            {
                name: 'Disenroll Selected',
                action: 'delete',
                danger: true,
            },
        ],
        filters: [
            {
                type: 'field_value_selection',
                field: 'type',
                name: 'Type',
                options: lodash_1.uniq(data.map(({ type }) => ({ value: type })), 'value'),
            },
        ],
    }),
};
exports.TagsTableType = {
    itemType: 'Tags',
    columnDefinitions: [
        {
            field: 'id',
            name: 'Tag name',
            render: (id, tag) => (react_1.default.createElement(connected_link_1.ConnectedLink, { path: `/tag/edit/${tag.id}` },
                react_1.default.createElement(tag_1.TagBadge, { tag: tag }))),
            sortable: true,
            width: '45%',
        },
        {
            align: 'right',
            field: 'configuration_blocks',
            name: 'Configurations',
            render: (configurationBlocks) => (react_1.default.createElement("div", null, configurationBlocks.length)),
            sortable: false,
        },
        {
            align: 'right',
            field: 'last_updated',
            name: 'Last update',
            render: (lastUpdate) => react_1.default.createElement("div", null, moment_1.default(lastUpdate).fromNow()),
            sortable: true,
        },
    ],
    controlDefinitions: (data) => ({
        actions: [
            {
                name: 'Remove Selected',
                action: 'delete',
                danger: true,
            },
        ],
        filters: [],
    }),
};
exports.BeatDetailTagsTable = {
    itemType: 'Tags',
    columnDefinitions: [
        {
            field: 'id',
            name: 'Tag name',
            render: (id, tag) => (react_1.default.createElement(connected_link_1.ConnectedLink, { path: `/tag/edit/${tag.id}` },
                react_1.default.createElement(tag_1.TagBadge, { tag: tag }))),
            sortable: true,
            width: '55%',
        },
        {
            align: 'right',
            field: 'configuration_blocks',
            name: 'Configurations',
            render: (configurations) => react_1.default.createElement("span", null, configurations.length),
            sortable: true,
        },
        {
            align: 'right',
            field: 'last_updated',
            name: 'Last update',
            render: (lastUpdate) => react_1.default.createElement("span", null, moment_1.default(lastUpdate).fromNow()),
            sortable: true,
        },
    ],
    controlDefinitions: (data) => ({
        actions: [],
        filters: [],
        primaryActions: [
            {
                name: 'Add Tag',
                action: 'add',
                danger: false,
            },
            {
                name: 'Remove Selected',
                action: 'remove',
                danger: true,
            },
        ],
    }),
};
