// x-pack/plugins/security/public/views/management/edit_role/components/privileges/kibana/space_aware_privilege_form.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importStar(require("react"));
const role_1 = require("../../../../../../lib/role");
const constants_1 = require("../../../lib/constants");
const copy_role_1 = require("../../../lib/copy_role");
const get_available_privileges_1 = require("../../../lib/get_available_privileges");
const impacted_spaces_flyout_1 = require("./impacted_spaces_flyout");
const privilege_callout_warning_1 = require("./privilege_callout_warning");
const privilege_selector_1 = require("./privilege_selector");
const privilege_space_form_1 = require("./privilege_space_form");
const privilege_space_table_1 = require("./privilege_space_table");
class SpaceAwarePrivilegeForm extends react_1.Component {
    constructor(props) {
        super(props);
        this.renderSpacePrivileges = (basePrivilege, availablePrivileges) => {
            const { role, spaces } = this.props;
            const { spacePrivileges } = this.state;
            const availableSpaces = this.getAvailableSpaces();
            const canAssignSpacePrivileges = basePrivilege !== 'all';
            const hasAssignedSpacePrivileges = Object.keys(this.state.spacePrivileges).length > 0;
            const showAddPrivilegeButton = canAssignSpacePrivileges && this.props.editable && availableSpaces.length > 0;
            return (react_1.default.createElement(react_1.Fragment, null,
                react_1.default.createElement(eui_1.EuiTitle, { size: 'xs' },
                    react_1.default.createElement("h3", null, "Higher privileges for individual spaces")),
                react_1.default.createElement(eui_1.EuiSpacer, { size: 's' }),
                react_1.default.createElement(eui_1.EuiText
                // @ts-ignore
                , { 
                    // @ts-ignore
                    grow: false, size: 's', color: 'subdued' },
                    react_1.default.createElement("p", null,
                        "Grant more privileges on a per space basis. For example, if the privileges are",
                        ' ',
                        react_1.default.createElement("strong", null, "read"),
                        " for all spaces, you can set the privileges to ",
                        react_1.default.createElement("strong", null, "all"),
                        ' ',
                        "for an individual space.")),
                react_1.default.createElement(eui_1.EuiSpacer, { size: 's' }),
                (basePrivilege !== constants_1.NO_PRIVILEGE_VALUE || role_1.isReservedRole(this.props.role)) && (react_1.default.createElement(privilege_callout_warning_1.PrivilegeCalloutWarning, { basePrivilege: basePrivilege, isReservedRole: role_1.isReservedRole(this.props.role) })),
                basePrivilege === 'read' && this.props.editable && react_1.default.createElement(eui_1.EuiSpacer, null),
                canAssignSpacePrivileges && (react_1.default.createElement(react_1.Fragment, null,
                    react_1.default.createElement(privilege_space_table_1.PrivilegeSpaceTable, { role: role, spaces: spaces, availablePrivileges: availablePrivileges, spacePrivileges: spacePrivileges, onChange: this.onExistingSpacePrivilegesChange }),
                    hasAssignedSpacePrivileges && react_1.default.createElement(eui_1.EuiSpacer, null),
                    this.getSpaceForms(basePrivilege))),
                react_1.default.createElement(eui_1.EuiFlexGroup
                // @ts-ignore
                , { 
                    // @ts-ignore
                    alignItems: 'baseline' },
                    showAddPrivilegeButton && (react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiButton, { "data-test-subj": "addSpacePrivilegeButton", size: 's', iconType: 'plusInCircle', onClick: this.addSpacePrivilege }, "Add space privilege"))),
                    react_1.default.createElement(eui_1.EuiFlexItem, null,
                        react_1.default.createElement(impacted_spaces_flyout_1.ImpactedSpacesFlyout, { role: role, spaces: spaces, userProfile: this.props.userProfile })))));
        };
        this.getSpaceForms = (basePrivilege) => {
            if (!this.props.editable) {
                return null;
            }
            return this.state.privilegeForms.map((form, index) => this.getSpaceForm(form, index, basePrivilege));
        };
        this.addSpacePrivilege = () => {
            this.setState({
                privilegeForms: [
                    ...this.state.privilegeForms,
                    {
                        spaces: [],
                        privilege: null,
                    },
                ],
            });
        };
        this.getAvailableSpaces = (omitIndex) => {
            const { spacePrivileges } = this.state;
            return this.props.spaces.filter(space => {
                const alreadyAssigned = Object.keys(spacePrivileges).indexOf(space.id) >= 0;
                if (alreadyAssigned) {
                    return false;
                }
                const otherForms = [...this.state.privilegeForms];
                if (typeof omitIndex === 'number') {
                    otherForms.splice(omitIndex, 1);
                }
                const inAnotherForm = otherForms.some(({ spaces }) => spaces.indexOf(space.id) >= 0);
                return !inAnotherForm;
            });
        };
        this.getSpaceForm = (form, index, basePrivilege) => {
            const { spaces: selectedSpaceIds, privilege } = form;
            const availableSpaces = this.getAvailableSpaces(index);
            return (react_1.default.createElement(react_1.Fragment, { key: index },
                react_1.default.createElement(privilege_space_form_1.PrivilegeSpaceForm, { key: index, availableSpaces: availableSpaces, selectedSpaceIds: selectedSpaceIds, availablePrivileges: get_available_privileges_1.getAvailablePrivileges(basePrivilege), selectedPrivilege: privilege, onChange: this.onPrivilegeSpacePermissionChange(index), onDelete: this.onPrivilegeSpacePermissionDelete(index), validator: this.props.validator }),
                react_1.default.createElement(eui_1.EuiSpacer, null)));
        };
        this.onPrivilegeSpacePermissionChange = (index) => (form) => {
            const existingPrivilegeForm = { ...this.state.privilegeForms[index] };
            const updatedPrivileges = [...this.state.privilegeForms];
            updatedPrivileges[index] = {
                spaces: form.spaces,
                privilege: form.privilege,
            };
            this.setState({
                privilegeForms: updatedPrivileges,
            });
            const role = copy_role_1.copyRole(this.props.role);
            existingPrivilegeForm.spaces.forEach(spaceId => {
                role.kibana.space[spaceId] = [];
            });
            if (form.spaces.length) {
                const privilege = form.privilege;
                if (privilege) {
                    form.spaces.forEach(spaceId => {
                        role.kibana.space[spaceId] = [privilege];
                    });
                }
            }
            this.props.validator.setInProgressSpacePrivileges(updatedPrivileges);
            this.props.onChange(role);
        };
        this.onPrivilegeSpacePermissionDelete = (index) => () => {
            const updatedPrivileges = [...this.state.privilegeForms];
            const removedPrivilege = updatedPrivileges.splice(index, 1)[0];
            this.setState({
                privilegeForms: updatedPrivileges,
            });
            const role = copy_role_1.copyRole(this.props.role);
            removedPrivilege.spaces.forEach(spaceId => {
                delete role.kibana.space[spaceId];
            });
            this.props.validator.setInProgressSpacePrivileges(updatedPrivileges);
            this.props.onChange(role);
        };
        this.onExistingSpacePrivilegesChange = (assignedPrivileges) => {
            const role = copy_role_1.copyRole(this.props.role);
            role.kibana.space = {
                ...assignedPrivileges,
            };
            // Merge in-progress forms
            this.state.privilegeForms.forEach(inProgressForm => {
                const { privilege } = inProgressForm;
                if (privilege) {
                    inProgressForm.spaces.forEach(spaceId => {
                        role.kibana.space[spaceId] = [privilege];
                    });
                }
            });
            this.setState({
                spacePrivileges: assignedPrivileges,
            });
            this.props.onChange(role);
        };
        this.onKibanaBasePrivilegeChange = (privilege) => {
            const role = copy_role_1.copyRole(this.props.role);
            // Remove base privilege value
            role.kibana.global = [];
            if (privilege !== constants_1.NO_PRIVILEGE_VALUE) {
                role.kibana.global = [privilege];
            }
            this.props.onChange(role);
        };
        const { role } = props;
        const assignedPrivileges = role.kibana;
        const spacePrivileges = {
            ...assignedPrivileges.space,
        };
        this.state = {
            spacePrivileges,
            privilegeForms: [],
        };
    }
    render() {
        const { kibanaAppPrivileges, role, userProfile } = this.props;
        if (!userProfile.hasCapability('manageSpaces')) {
            return (react_1.default.createElement(eui_1.EuiCallOut, { title: react_1.default.createElement("p", null, "Insufficient Privileges"), iconType: "alert", color: "danger" },
                react_1.default.createElement("p", null, "You are not authorized to view all available spaces."),
                react_1.default.createElement("p", null,
                    "Please ensure your account has all privileges granted by the",
                    ' ',
                    react_1.default.createElement("strong", null, "kibana_user"),
                    " role, and try again.")));
        }
        const assignedPrivileges = role.kibana;
        const basePrivilege = assignedPrivileges.global.length > 0 ? assignedPrivileges.global[0] : constants_1.NO_PRIVILEGE_VALUE;
        const description = react_1.default.createElement("p", null, "Specify the minimum actions users can perform in your spaces.");
        let helptext;
        if (basePrivilege === constants_1.NO_PRIVILEGE_VALUE) {
            helptext = 'No access to spaces';
        }
        else if (basePrivilege === 'all') {
            helptext = 'View, edit, and share objects and apps within all spaces';
        }
        else if (basePrivilege === 'read') {
            helptext = 'View objects and apps within all spaces';
        }
        return (react_1.default.createElement(react_1.Fragment, null,
            react_1.default.createElement(eui_1.EuiDescribedFormGroup, { title: react_1.default.createElement("h3", null, "Minimum privileges for all spaces"), description: description },
                react_1.default.createElement(eui_1.EuiFormRow, { hasEmptyLabelSpace: true, helpText: helptext },
                    react_1.default.createElement(privilege_selector_1.PrivilegeSelector, { "data-test-subj": 'kibanaMinimumPrivilege', availablePrivileges: kibanaAppPrivileges, value: basePrivilege, disabled: role_1.isReservedRole(role), allowNone: true, onChange: this.onKibanaBasePrivilegeChange }))),
            react_1.default.createElement(eui_1.EuiSpacer, null),
            this.renderSpacePrivileges(basePrivilege, kibanaAppPrivileges)));
    }
}
exports.SpaceAwarePrivilegeForm = SpaceAwarePrivilegeForm;
