// x-pack/plugins/beats_management/public/components/tag/tag_edit.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
require("brace/mode/yaml");
require("brace/theme/github");
const lodash_1 = require("lodash");
const react_1 = tslib_1.__importDefault(require("react"));
const config_list_1 = require("../config_list");
const table_1 = require("../table");
const table_2 = require("../table");
const table_3 = require("../table");
const config_view_1 = require("./config_view");
const tag_badge_1 = require("./tag_badge");
class TagEdit extends react_1.default.PureComponent {
    constructor(props) {
        super(props);
        this.getNameError = (name) => {
            if (name && name !== '' && name.search(/^[a-zA-Z0-9-]+$/) === -1) {
                return 'Tag name must consist of letters, numbers, and dashes only';
            }
            else {
                return false;
            }
        };
        this.handleAssignmentActions = (action) => {
            switch (action) {
                case table_1.AssignmentActionType.Delete:
                    const { selection } = this.state.tableRef.current.state;
                    this.props.onDetachBeat(selection.map((beat) => beat.id));
            }
        };
        // TODO this should disable save button on bad validations
        this.updateTag = (key, value) => value !== undefined
            ? this.props.onTagChange(key, value)
            : (e) => this.props.onTagChange(key, e.target ? e.target.value : e);
        this.state = {
            showFlyout: false,
            tableRef: react_1.default.createRef(),
        };
    }
    render() {
        const { tag, attachedBeats } = this.props;
        return (react_1.default.createElement("div", null,
            react_1.default.createElement(eui_1.EuiFlexGroup, null,
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement(eui_1.EuiTitle, { size: "xs" },
                        react_1.default.createElement("h3", null, "Tag details")),
                    react_1.default.createElement(eui_1.EuiText, { color: "subdued" },
                        react_1.default.createElement("p", null, "A tag is a group of configuration blocks that you can apply to one or more Beats.")),
                    react_1.default.createElement("div", null,
                        react_1.default.createElement(tag_badge_1.TagBadge, { tag: { color: tag.color || '#FF0', id: tag.id } }))),
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement(eui_1.EuiForm, null,
                        react_1.default.createElement(eui_1.EuiFormRow, { label: "Tag Name", isInvalid: !!this.getNameError(tag.id), error: this.getNameError(tag.id) || undefined },
                            react_1.default.createElement(eui_1.EuiFieldText, { name: "name", isInvalid: !!this.getNameError(tag.id), onChange: this.updateTag('id'), disabled: this.props.mode === 'edit', value: tag.id, placeholder: "Tag name (required)" })),
                        this.props.mode === 'create' && (react_1.default.createElement(eui_1.EuiFormRow, { label: "Tag Color" },
                            react_1.default.createElement(eui_1.EuiColorPicker, { color: tag.color, onChange: this.updateTag('color') })))))),
            react_1.default.createElement(eui_1.EuiSpacer, null),
            react_1.default.createElement(eui_1.EuiHorizontalRule, null),
            react_1.default.createElement(eui_1.EuiFlexGroup, { alignItems: tag.configuration_blocks && tag.configuration_blocks.length ? 'stretch' : 'center' },
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement(eui_1.EuiTitle, { size: "xs" },
                        react_1.default.createElement("h3", null, "Configuration blocks")),
                    react_1.default.createElement(eui_1.EuiText, { color: "subdued" },
                        react_1.default.createElement("p", null, "A tag can have configuration blocks for different types of Beats. For example, a tag can have two Metricbeat configuration blocks and one Filebeat input configuration block."))),
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement("div", null,
                        react_1.default.createElement(config_list_1.ConfigList, { configs: tag.configuration_blocks, onConfigClick: (action, config) => {
                                const selectedIndex = tag.configuration_blocks.findIndex(c => {
                                    return lodash_1.isEqual(config, c);
                                });
                                if (action === 'delete') {
                                    const configs = [...tag.configuration_blocks];
                                    configs.splice(selectedIndex, 1);
                                    this.updateTag('configuration_blocks', configs);
                                }
                                else {
                                    this.setState({
                                        showFlyout: true,
                                        selectedConfigIndex: selectedIndex,
                                    });
                                }
                            } }),
                        react_1.default.createElement("br", null),
                        react_1.default.createElement(eui_1.EuiButton, { onClick: () => {
                                this.setState({ showFlyout: true });
                            } }, "Add configuration block")))),
            react_1.default.createElement(eui_1.EuiSpacer, null),
            attachedBeats && (react_1.default.createElement("div", null,
                react_1.default.createElement(eui_1.EuiHorizontalRule, null),
                react_1.default.createElement(eui_1.EuiTitle, { size: "xs" },
                    react_1.default.createElement("h3", null, "Beats with this tag")),
                react_1.default.createElement(table_1.Table, { assignmentOptions: {
                        schema: table_3.tagConfigAssignmentOptions,
                        items: [],
                        type: 'primary',
                        actionHandler: this.handleAssignmentActions,
                    }, items: attachedBeats, ref: this.state.tableRef, type: table_2.BeatsTableType }))),
            this.state.showFlyout && (react_1.default.createElement(config_view_1.ConfigView, { configBlock: this.state.selectedConfigIndex !== undefined
                    ? tag.configuration_blocks[this.state.selectedConfigIndex]
                    : undefined, onClose: () => this.setState({ showFlyout: false, selectedConfigIndex: undefined }), onSave: (config) => {
                    this.setState({ showFlyout: false, selectedConfigIndex: undefined });
                    if (this.state.selectedConfigIndex !== undefined) {
                        const configs = [...tag.configuration_blocks];
                        configs[this.state.selectedConfigIndex] = config;
                        this.updateTag('configuration_blocks', configs);
                    }
                    else {
                        this.updateTag('configuration_blocks', [
                            ...(tag.configuration_blocks || []),
                            config,
                        ]);
                    }
                } }))));
    }
}
exports.TagEdit = TagEdit;
