// x-pack/plugins/infra/public/components/logging/log_time_controls.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const moment_1 = tslib_1.__importDefault(require("moment"));
const react_1 = tslib_1.__importDefault(require("react"));
const styled_components_1 = tslib_1.__importDefault(require("styled-components"));
const noop = () => undefined;
class LogTimeControls extends react_1.default.PureComponent {
    constructor() {
        super(...arguments);
        this.handleChangeDate = (date) => {
            if (date !== null) {
                this.props.jumpToTime(date.valueOf());
            }
        };
        this.startLiveStreaming = () => {
            this.props.startLiveStreaming(5000);
        };
        this.stopLiveStreaming = () => {
            this.props.stopLiveStreaming();
        };
    }
    render() {
        const { currentTime, isLiveStreaming } = this.props;
        const currentMoment = currentTime ? moment_1.default(currentTime) : null;
        if (isLiveStreaming) {
            return (react_1.default.createElement(eui_1.EuiFilterGroup, null,
                react_1.default.createElement(InlineWrapper, null,
                    react_1.default.createElement(eui_1.EuiDatePicker, { disabled: true, onChange: noop, value: "streaming..." })),
                react_1.default.createElement(eui_1.EuiFilterButton, { color: "primary", iconType: "pause", iconSide: "left", onClick: this.stopLiveStreaming }, "Stop streaming")));
        }
        else {
            return (react_1.default.createElement(eui_1.EuiFilterGroup, null,
                react_1.default.createElement(InlineWrapper, null,
                    react_1.default.createElement(eui_1.EuiDatePicker, { dateFormat: "L LTS", onChange: this.handleChangeDate, popperPlacement: "top-end", selected: currentMoment, shouldCloseOnSelect: true, showTimeSelect: true, timeFormat: "LTS", injectTimes: currentMoment ? [currentMoment] : [] })),
                react_1.default.createElement(eui_1.EuiFilterButton, { iconType: "play", iconSide: "left", onClick: this.startLiveStreaming }, "Stream live")));
        }
    }
}
exports.LogTimeControls = LogTimeControls;
const InlineWrapper = styled_components_1.default.div `
  display: inline-block;
`;
