// @elastic/eui/types/components/spacer/spacer.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
import React from 'react';
import classNames from 'classnames';
var sizeToClassNameMap = {
    xs: 'euiSpacer--xs',
    s: 'euiSpacer--s',
    m: 'euiSpacer--m',
    l: 'euiSpacer--l',
    xl: 'euiSpacer--xl',
    xxl: 'euiSpacer--xxl',
};
export var SIZES = Object.keys(sizeToClassNameMap);
export var EuiSpacer = function (_a) {
    var className = _a.className, size = _a.size, rest = __rest(_a, ["className", "size"]);
    var classes = classNames('euiSpacer', size ? sizeToClassNameMap[size] : undefined, className);
    return (React.createElement("div", __assign({ className: classes }, rest)));
};
EuiSpacer.defaultProps = {
    size: 'l',
};
