// x-pack/plugins/beats_management/public/components/tag/config_view/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importDefault(require("react"));
const config_schemas_1 = require("../../../config_schemas");
const config_form_1 = require("./config_form");
class ConfigView extends react_1.default.Component {
    constructor(props) {
        super(props);
        this.form = react_1.default.createRef();
        this.onValueChange = (field) => (e) => {
            const value = e.currentTarget ? e.currentTarget.value : e;
            this.setState((state) => ({
                configBlock: {
                    ...state.configBlock,
                    [field]: value,
                },
            }));
        };
        this.editMode = props.configBlock !== undefined;
        this.state = {
            valid: false,
            configBlock: props.configBlock || {
                type: config_schemas_1.supportedConfigs[0].value,
            },
        };
    }
    render() {
        return (react_1.default.createElement(eui_1.EuiFlyout, { onClose: this.props.onClose },
            react_1.default.createElement(eui_1.EuiFlyoutHeader, null,
                react_1.default.createElement(eui_1.EuiTitle, { size: "m" },
                    react_1.default.createElement("h2", null, this.editMode
                        ? this.props.onSave
                            ? 'Edit configuration block'
                            : 'View configuration block'
                        : 'Add configuration block'))),
            react_1.default.createElement(eui_1.EuiFlyoutBody, null,
                react_1.default.createElement(eui_1.EuiFormRow, { label: "Type" },
                    react_1.default.createElement(eui_1.EuiSelect, { options: config_schemas_1.supportedConfigs, value: this.state.configBlock.type, disabled: this.editMode, onChange: this.onValueChange('type') })),
                react_1.default.createElement(eui_1.EuiFormRow, { label: "Description" },
                    react_1.default.createElement(eui_1.EuiFieldText, { value: this.state.configBlock.description, disabled: !this.props.onSave, onChange: this.onValueChange('description'), placeholder: "Description (optional)" })),
                react_1.default.createElement("h3", null,
                    config_schemas_1.supportedConfigs.find(config => this.state.configBlock.type === config.value)
                        .text,
                    "\u00A0configuration"),
                react_1.default.createElement(eui_1.EuiHorizontalRule, null),
                react_1.default.createElement(config_form_1.ConfigForm, { onSubmit: this.props.onSave
                        ? data => {
                            if (this.props.onSave) {
                                this.props.onSave({
                                    ...this.state.configBlock,
                                    configs: [data],
                                });
                            }
                            this.props.onClose();
                        }
                        : undefined, canSubmit: canIt => this.setState({ valid: canIt }), ref: this.form, values: this.state.configBlock, id: config_schemas_1.supportedConfigs.find(config => this.state.configBlock.type === config.value)
                        .value, schema: config_schemas_1.supportedConfigs.find(config => this.state.configBlock.type === config.value)
                        .config })),
            react_1.default.createElement(eui_1.EuiFlyoutFooter, null,
                react_1.default.createElement(eui_1.EuiFlexGroup, { justifyContent: "spaceBetween" },
                    react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiButtonEmpty, { iconType: "cross", onClick: this.props.onClose }, "Close")),
                    this.props.onSave && (react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiButton, { disabled: !this.state.valid, fill: true, onClick: () => {
                                if (this.form.current) {
                                    this.form.current.submit();
                                }
                            } }, "Save")))))));
    }
}
exports.ConfigView = ConfigView;
