// x-pack/plugins/apm/public/components/app/Main/routeConfig.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const react_1 = tslib_1.__importDefault(require("react"));
const react_router_dom_1 = require("react-router-dom");
const url_1 = require("../../../utils/url");
// @ts-ignore
const ErrorGroupDetails_1 = tslib_1.__importDefault(require("../ErrorGroupDetails"));
// @ts-ignore
const ErrorGroupOverview_1 = tslib_1.__importDefault(require("../ErrorGroupOverview"));
const TransactionDetails_1 = require("../TransactionDetails");
// @ts-ignore
const TransactionOverview_1 = tslib_1.__importDefault(require("../TransactionOverview"));
const Home_1 = require("./Home");
exports.routes = [
    {
        exact: true,
        path: '/',
        component: Home_1.Home,
        breadcrumb: 'APM'
    },
    {
        exact: true,
        path: '/:serviceName/errors/:groupId',
        component: ErrorGroupDetails_1.default,
        breadcrumb: ({ match }) => match.params.groupId
    },
    {
        exact: true,
        path: '/:serviceName/errors',
        component: ErrorGroupOverview_1.default,
        breadcrumb: 'Errors'
    },
    {
        switch: true,
        routes: [
            {
                exact: true,
                path: '/invalid-license',
                breadcrumb: 'Invalid License',
                render: () => react_1.default.createElement("div", null, "Invalid license")
            },
            {
                exact: true,
                path: '/:serviceName',
                breadcrumb: ({ match }) => match.params.serviceName,
                render: ({ location }) => {
                    return (react_1.default.createElement(react_router_dom_1.Redirect, { to: {
                            ...location,
                            pathname: `${location.pathname}/transactions`
                        } }));
                }
            }
        ]
    },
    {
        exact: true,
        path: '/:serviceName/transactions',
        component: TransactionOverview_1.default,
        breadcrumb: 'Transactions'
    },
    {
        exact: true,
        path: '/:serviceName/transactions/:transactionType',
        component: TransactionOverview_1.default,
        breadcrumb: ({ match }) => url_1.legacyDecodeURIComponent(match.params.transactionType)
    },
    {
        exact: true,
        path: '/:serviceName/transactions/:transactionType/:transactionName',
        component: TransactionDetails_1.TransactionDetails,
        breadcrumb: ({ match }) => url_1.legacyDecodeURIComponent(match.params.transactionName)
    }
];
