// x-pack/plugins/beats_management/public/pages/beat/detail.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const lodash_1 = require("lodash");
const react_1 = tslib_1.__importDefault(require("react"));
const constants_1 = require("../../../common/constants");
const connected_link_1 = require("../../components/connected_link");
const tag_1 = require("../../components/tag");
const index_1 = require("../../components/tag/config_view/index");
const config_schemas_1 = require("../../config_schemas");
class BeatDetailPage extends react_1.default.PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            selectedConfig: null,
        };
    }
    render() {
        const props = this.props;
        const { beat } = props;
        if (!beat) {
            return react_1.default.createElement("div", null, "Beat not found");
        }
        const configurationBlocks = lodash_1.flatten(beat.full_tags.map((tag) => {
            return tag.configuration_blocks.map(configuration => ({
                // @ts-ignore one of the types on ConfigurationBlock doesn't define a "module" property
                module: configuration.configs[0].module || null,
                tagId: tag.id,
                tagColor: tag.color,
                ...beat,
                ...configuration,
                displayValue: lodash_1.get(config_schemas_1.supportedConfigs.find(config => config.value === configuration.type), 'text', null),
            }));
        }));
        const columns = [
            {
                field: 'displayValue',
                name: 'Type',
                sortable: true,
                render: (value, configuration) => (react_1.default.createElement(eui_1.EuiLink, { onClick: () => {
                        this.setState({
                            selectedConfig: configuration,
                        });
                    } }, value || configuration.type)),
            },
            {
                field: 'module',
                name: 'Module',
                sortable: true,
            },
            {
                field: 'description',
                name: 'Description',
                sortable: true,
            },
            {
                field: 'tagId',
                name: 'Tag',
                render: (id, block) => (react_1.default.createElement(connected_link_1.ConnectedLink, { path: `/tag/edit/${id}` },
                    react_1.default.createElement(tag_1.TagBadge, { maxIdRenderSize: constants_1.TABLE_CONFIG.TRUNCATE_TAG_LENGTH_SMALL, tag: { color: block.tagColor, id } }))),
                sortable: true,
            },
        ];
        return (react_1.default.createElement(react_1.default.Fragment, null,
            react_1.default.createElement(eui_1.EuiFlexGroup, null,
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement(eui_1.EuiTitle, { size: "xs" },
                        react_1.default.createElement("h4", null, "Configurations")),
                    react_1.default.createElement(eui_1.EuiText, { size: "s" },
                        react_1.default.createElement("p", null, "You can have multiple configurations applied to an individual tag. These configurations can repeat or mix types as necessary. For example, you may utilize three metricbeat configurations alongside one input and filebeat configuration."))),
                react_1.default.createElement(eui_1.EuiFlexItem, null,
                    react_1.default.createElement(eui_1.EuiInMemoryTable, { columns: columns, items: configurationBlocks }))),
            this.state.selectedConfig && (react_1.default.createElement(index_1.ConfigView, { configBlock: this.state.selectedConfig, onClose: () => this.setState({ selectedConfig: null }) }))));
    }
}
exports.BeatDetailPage = BeatDetailPage;
