// x-pack/plugins/beats_management/public/components/config_list.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
// @ts-ignore
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importDefault(require("react"));
const config_schemas_1 = require("../config_schemas");
exports.ConfigList = props => (react_1.default.createElement(eui_1.EuiBasicTable, { items: props.configs || [], columns: [
        {
            field: 'type',
            name: 'Type',
            truncateText: false,
            render: (value, config) => {
                const type = config_schemas_1.supportedConfigs.find((sc) => sc.value === config.type);
                return (react_1.default.createElement(eui_1.EuiLink, { onClick: () => props.onConfigClick('edit', config) }, type ? type.text : config.type));
            },
        },
        {
            field: 'module',
            name: 'Module',
            truncateText: false,
            render: (value) => {
                return value || 'N/A';
            },
        },
        {
            field: 'description',
            name: 'Description',
        },
        {
            name: 'Actions',
            actions: [
                {
                    name: 'Remove',
                    description: 'Remove this config from tag',
                    type: 'icon',
                    icon: 'trash',
                    onClick: (item) => props.onConfigClick('delete', item),
                },
            ],
        },
    ] }));
