// x-pack/plugins/reporting/public/components/reporting_panel_content.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importStar(require("react"));
const notify_1 = require("ui/notify");
const url_1 = tslib_1.__importDefault(require("url"));
const reporting_client_1 = require("../lib/reporting_client");
class ReportingPanelContent extends react_1.Component {
    constructor(props) {
        super(props);
        this.renderGenerateReportButton = (isDisabled) => {
            return (react_1.default.createElement(eui_1.EuiButton, { className: "sharePanel__button", disabled: isDisabled, fill: true, onClick: this.createReportingJob, "data-test-subj": "generateReportButton", size: "s" },
                "Generate ",
                this.prettyPrintReportingType()));
        };
        this.prettyPrintReportingType = () => {
            switch (this.props.reportType) {
                case 'printablePdf':
                    return 'PDF';
                case 'csv':
                    return 'CSV';
                default:
                    return this.props.reportType;
            }
        };
        this.markAsStale = () => {
            if (!this.mounted) {
                return;
            }
            this.setState({ isStale: true });
        };
        this.isNotSaved = () => {
            return this.props.objectId === undefined || this.props.objectId === '';
        };
        this.setAbsoluteReportGenerationUrl = () => {
            if (!this.mounted) {
                return;
            }
            const relativePath = reporting_client_1.reportingClient.getReportingJobPath(this.props.reportType, this.props.getJobParams());
            const absoluteUrl = url_1.default.resolve(window.location.href, relativePath);
            this.setState({ absoluteUrl });
        };
        this.createReportingJob = () => {
            return reporting_client_1.reportingClient
                .createReportingJob(this.props.reportType, this.props.getJobParams())
                .then(() => {
                notify_1.toastNotifications.addSuccess({
                    title: `Queued report for ${this.props.objectType}`,
                    text: 'Track its progress in Management',
                    'data-test-subj': 'queueReportSuccess',
                });
                this.props.onClose();
            })
                .catch((kfetchError) => {
                if (kfetchError.message === 'not exportable') {
                    return notify_1.toastNotifications.addWarning({
                        title: `Only saved ${this.props.objectType} can be exported`,
                        text: 'Please save your work first',
                    });
                }
                const defaultMessage = kfetchError.res.status === 403
                    ? `You don't have permission to generate this report.`
                    : `Can't reach the server. Please try again.`;
                notify_1.toastNotifications.addDanger({
                    title: 'Reporting error',
                    text: kfetchError.message || defaultMessage,
                    'data-test-subj': 'queueReportError',
                });
            });
        };
        this.state = {
            isStale: false,
            absoluteUrl: '',
        };
    }
    componentWillUnmount() {
        window.removeEventListener('hashchange', this.markAsStale);
        window.removeEventListener('resize', this.setAbsoluteReportGenerationUrl);
        this.mounted = false;
    }
    componentDidMount() {
        this.mounted = true;
        this.setAbsoluteReportGenerationUrl();
        window.addEventListener('hashchange', this.markAsStale, false);
        window.addEventListener('resize', this.setAbsoluteReportGenerationUrl);
    }
    render() {
        if (this.isNotSaved() || this.props.isDirty || this.state.isStale) {
            return (react_1.default.createElement(eui_1.EuiForm, { className: "sharePanelContent", "data-test-subj": "shareReportingForm" },
                react_1.default.createElement(eui_1.EuiFormRow, { helpText: 'Please save your work before generating a report.' }, this.renderGenerateReportButton(true))));
        }
        const reportMsg = `${this.prettyPrintReportingType()}s can take a minute or two to generate based upon the size of your ${this.props.objectType}.`;
        return (react_1.default.createElement(eui_1.EuiForm, { className: "sharePanelContent", "data-test-subj": "shareReportingForm" },
            react_1.default.createElement(eui_1.EuiText, { size: "s" },
                react_1.default.createElement("p", null, reportMsg)),
            react_1.default.createElement(eui_1.EuiSpacer, { size: "s" }),
            this.props.options,
            this.renderGenerateReportButton(false),
            react_1.default.createElement(eui_1.EuiSpacer, { size: "s" }),
            react_1.default.createElement(eui_1.EuiText, { size: "s" },
                react_1.default.createElement("p", null, "Alternatively, copy this POST URL to call generation from outside Kibana or from Watcher.")),
            react_1.default.createElement(eui_1.EuiSpacer, { size: "s" }),
            react_1.default.createElement(eui_1.EuiCopy, { textToCopy: this.state.absoluteUrl, anchorClassName: "sharePanel__copyAnchor" }, (copy) => (react_1.default.createElement(eui_1.EuiButton, { className: "sharePanel__button", onClick: copy, size: "s" }, "Copy POST URL")))));
    }
}
exports.ReportingPanelContent = ReportingPanelContent;
