// x-pack/plugins/reporting/public/components/screen_capture_panel_content.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const eui_1 = require("@elastic/eui");
const react_1 = tslib_1.__importStar(require("react"));
const reporting_panel_content_1 = require("./reporting_panel_content");
class ScreenCapturePanelContent extends react_1.Component {
    constructor(props) {
        super(props);
        this.renderOptions = () => {
            return (react_1.default.createElement(react_1.Fragment, null,
                react_1.default.createElement(eui_1.EuiSwitch, { label: "Optimize for printing", checked: this.state.usePrintLayout, onChange: this.handlePrintLayoutChange, "data-test-subj": "usePrintLayout" }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "s" })));
        };
        this.handlePrintLayoutChange = (evt) => {
            this.setState({ usePrintLayout: evt.target.checked });
        };
        this.getLayout = () => {
            if (this.state.usePrintLayout) {
                return { id: 'print' };
            }
            const el = document.querySelector('[data-shared-items-container]');
            const bounds = el ? el.getBoundingClientRect() : { height: 768, width: 1024 };
            return {
                id: 'preserve_layout',
                dimensions: {
                    height: bounds.height,
                    width: bounds.width,
                },
            };
        };
        this.getJobParams = () => {
            const jobParams = this.props.getJobParams();
            jobParams.layout = this.getLayout();
            return jobParams;
        };
        this.state = {
            usePrintLayout: false,
        };
    }
    render() {
        return (react_1.default.createElement(reporting_panel_content_1.ReportingPanelContent, { reportType: this.props.reportType, objectType: this.props.objectType, objectId: this.props.objectId, getJobParams: this.getJobParams, options: this.renderOptions(), isDirty: this.props.isDirty, onClose: this.props.onClose }));
    }
}
exports.ScreenCapturePanelContent = ScreenCapturePanelContent;
