// x-pack/plugins/apm/server/lib/helpers/transaction_group_query.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const moment_1 = tslib_1.__importDefault(require("moment"));
const constants_1 = require("../../../common/constants");
exports.TRANSACTION_GROUP_AGGREGATES = {
    transactions: {
        terms: {
            field: `${constants_1.TRANSACTION_NAME}.keyword`,
            order: { avg: 'desc' },
            size: 100
        },
        aggs: {
            sample: {
                top_hits: {
                    size: 1,
                    sort: [{ '@timestamp': { order: 'desc' } }]
                }
            },
            avg: { avg: { field: constants_1.TRANSACTION_DURATION } },
            p95: { percentiles: { field: constants_1.TRANSACTION_DURATION, percents: [95] } }
        }
    }
};
function calculateRelativeImpacts(results) {
    const values = results.map(({ impact }) => impact);
    const max = Math.max(...values);
    const min = Math.min(...values);
    return results.map(bucket => ({
        ...bucket,
        impact: ((bucket.impact - min) / (max - min)) * 100
    }));
}
function prepareTransactionGroups({ buckets, start, end }) {
    const duration = moment_1.default.duration(end - start);
    const minutes = duration.asMinutes();
    const results = buckets.map((bucket) => {
        const averageResponseTime = bucket.avg.value;
        const transactionsPerMinute = bucket.doc_count / minutes;
        const impact = Math.round(averageResponseTime * transactionsPerMinute);
        const sample = bucket.sample.hits.hits[0]._source;
        return {
            name: bucket.key,
            sample,
            p95: bucket.p95.values['95.0'],
            averageResponseTime,
            transactionsPerMinute,
            impact
        };
    });
    return calculateRelativeImpacts(results);
}
exports.prepareTransactionGroups = prepareTransactionGroups;
