// x-pack/plugins/infra/public/pages/metrics/layouts/container.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../../../common/graphql/types");
const lib_1 = require("../../../lib/lib");
const nginx_1 = require("./nginx");
const types_2 = require("./types");
exports.containerLayoutCreator = theme => [
    {
        id: 'containerOverview',
        label: 'Container Overview',
        sections: [
            {
                id: types_1.InfraMetric.containerOverview,
                label: 'Overview',
                requires: ['docker.cpu', 'docker.memory', 'docker.network'],
                type: types_2.InfraMetricLayoutSectionType.gauges,
                visConfig: {
                    seriesOverrides: {
                        cpu: {
                            name: 'CPU Usage',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.percent,
                            gaugeMax: 1,
                        },
                        memory: {
                            name: 'Memory Usage',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.percent,
                            gaugeMax: 1,
                        },
                        rx: {
                            name: 'Inbound (RX)',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.bits,
                            formatterTemplate: '{{value}}/s',
                        },
                        tx: {
                            name: 'Outbound (RX)',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.bits,
                            formatterTemplate: '{{value}}/s',
                        },
                    },
                },
            },
            {
                id: types_1.InfraMetric.containerCpuUsage,
                label: 'CPU Usage',
                requires: ['docker.cpu'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    stacked: true,
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    formatter: lib_1.InfraFormatterType.percent,
                    seriesOverrides: {
                        cpu: { color: theme.eui.euiColorVis1 },
                    },
                },
            },
            {
                id: types_1.InfraMetric.containerMemory,
                label: 'Memory Usage',
                requires: ['docker.memory'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    stacked: true,
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    formatter: lib_1.InfraFormatterType.percent,
                    seriesOverrides: {
                        memory: { color: theme.eui.euiColorVis1 },
                    },
                },
            },
            {
                id: types_1.InfraMetric.containerNetworkTraffic,
                label: 'Network Traffic',
                requires: ['docker.network'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.bits,
                    formatterTemplate: '{{value}}/s',
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    seriesOverrides: {
                        rx: { color: theme.eui.euiColorVis1, name: 'in' },
                        tx: { color: theme.eui.euiColorVis2, name: 'out' },
                    },
                },
            },
            {
                id: types_1.InfraMetric.containerDiskIOOps,
                label: 'Disk IO (Ops)',
                requires: ['docker.diskio'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.number,
                    formatterTemplate: '{{value}}/s',
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    seriesOverrides: {
                        read: { color: theme.eui.euiColorVis1, name: 'reads' },
                        write: { color: theme.eui.euiColorVis2, name: 'writes' },
                    },
                },
            },
            {
                id: types_1.InfraMetric.containerDiskIOBytes,
                label: 'Disk IO (Bytes)',
                requires: ['docker.diskio'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.bytes,
                    formatterTemplate: '{{value}}/s',
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    seriesOverrides: {
                        read: { color: theme.eui.euiColorVis1, name: 'reads' },
                        write: { color: theme.eui.euiColorVis2, name: 'writes' },
                    },
                },
            },
        ],
    },
    ...nginx_1.nginxLayoutCreator(theme),
];
