// x-pack/plugins/infra/public/pages/metrics/layouts/pod.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("../../../../common/graphql/types");
const lib_1 = require("../../../lib/lib");
const nginx_1 = require("./nginx");
const types_2 = require("./types");
exports.podLayoutCreator = theme => [
    {
        id: 'podOverview',
        label: 'Pod Overview',
        sections: [
            {
                id: types_1.InfraMetric.podOverview,
                label: 'Pod Overview',
                requires: ['kubernetes.pod'],
                type: types_2.InfraMetricLayoutSectionType.gauges,
                visConfig: {
                    seriesOverrides: {
                        cpu: {
                            name: 'CPU Usage',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.percent,
                            gaugeMax: 1,
                        },
                        memory: {
                            name: 'Memory Usage',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.percent,
                            gaugeMax: 1,
                        },
                        rx: {
                            name: 'Inbound (RX)',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.bits,
                            formatterTemplate: '{{value}}/s',
                        },
                        tx: {
                            name: 'Outbound (RX)',
                            color: theme.eui.euiColorFullShade,
                            formatter: lib_1.InfraFormatterType.bits,
                            formatterTemplate: '{{value}}/s',
                        },
                    },
                },
            },
            {
                id: types_1.InfraMetric.podCpuUsage,
                label: 'CPU Usage',
                requires: ['kubernetes.pod'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.percent,
                    seriesOverrides: {
                        cpu: { color: theme.eui.euiColorVis1, type: types_2.InfraMetricLayoutVisualizationType.area },
                    },
                },
            },
            {
                id: types_1.InfraMetric.podMemoryUsage,
                label: 'Memory Usage',
                requires: ['kubernetes.pod'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.percent,
                    seriesOverrides: {
                        memory: {
                            color: theme.eui.euiColorVis1,
                            type: types_2.InfraMetricLayoutVisualizationType.area,
                        },
                    },
                },
            },
            {
                id: types_1.InfraMetric.podNetworkTraffic,
                label: 'Network Traffic',
                requires: ['kubernetes.pod'],
                type: types_2.InfraMetricLayoutSectionType.chart,
                visConfig: {
                    formatter: lib_1.InfraFormatterType.bits,
                    formatterTemplate: '{{value}}/s',
                    type: types_2.InfraMetricLayoutVisualizationType.area,
                    seriesOverrides: {
                        rx: { color: theme.eui.euiColorVis1, name: 'in' },
                        tx: { color: theme.eui.euiColorVis2, name: 'out' },
                    },
                },
            },
        ],
    },
    ...nginx_1.nginxLayoutCreator(theme),
];
