// @elastic/eui/es/components/avatar/avatar.js
import "core-js/modules/es6.object.assign";
import "core-js/modules/es6.string.iterator";
import "core-js/modules/es6.array.from";
import "core-js/modules/es6.regexp.to-string";
import "core-js/modules/es7.symbol.async-iterator";
import "core-js/modules/es6.symbol";
import "core-js/modules/es6.regexp.match";
import "core-js/modules/es6.regexp.split";
import "core-js/modules/es6.function.name";
import "core-js/modules/web.dom.iterable";
import "core-js/modules/es6.array.iterator";
import "core-js/modules/es6.object.keys";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { isColorDark, hexToRgb } from '../../services/color';
import { VISUALIZATION_COLORS } from '../../services';
var sizeToClassNameMap = {
  'none': null,
  's': 'euiAvatar--s',
  'm': 'euiAvatar--m',
  'l': 'euiAvatar--l',
  'xl': 'euiAvatar--xl'
};
export var SIZES = Object.keys(sizeToClassNameMap);
var typeToClassNameMap = {
  space: 'euiAvatar--space',
  user: 'euiAvatar--user'
};
export var TYPES = Object.keys(typeToClassNameMap);
export var EuiAvatar = function EuiAvatar(_ref) {
  var className = _ref.className,
      color = _ref.color,
      imageUrl = _ref.imageUrl,
      initials = _ref.initials,
      initialsLength = _ref.initialsLength,
      name = _ref.name,
      size = _ref.size,
      type = _ref.type,
      rest = _objectWithoutProperties(_ref, ["className", "color", "imageUrl", "initials", "initialsLength", "name", "size", "type"]);

  var classes = classNames('euiAvatar', sizeToClassNameMap[size], typeToClassNameMap[type], className);
  var optionalInitial;

  if (name && !imageUrl) {
    // Calculate the number of initials to show, maxing out at 2
    var calculatedInitialsLength = initials ? initials.split(' ').length : name.split(' ').length;
    calculatedInitialsLength = calculatedInitialsLength > 2 ? 2 : calculatedInitialsLength; // Check if initialsLength was passed and set to calculated, unless greater than 2

    if (initialsLength) {
      calculatedInitialsLength = initialsLength <= 2 ? initialsLength : 2;
    }

    var calculatedInitials; // A. Set to initials prop if exists (but trancate to 2 characters max unless length is supplied)

    if (initials) {
      calculatedInitials = initials.substring(0, calculatedInitialsLength);
    } else {
      if (name.trim() && name.split(' ').length > 1) {
        // B. If there are any spaces in the name, set to first letter of each word
        calculatedInitials = name.match(/\b(\w)/g).join('').substring(0, calculatedInitialsLength);
      } else {
        // C. Set to the name's initials truncated based on calculated length
        calculatedInitials = name.substring(0, calculatedInitialsLength);
      }
    }

    optionalInitial = React.createElement("span", {
      "aria-hidden": "true"
    }, calculatedInitials);
  }

  var assignedColor = color || VISUALIZATION_COLORS[Math.floor(name.length % VISUALIZATION_COLORS.length)];
  var textColor = isColorDark.apply(void 0, _toConsumableArray(hexToRgb(assignedColor))) ? '#FFFFFF' : '#000000';
  var avatarStyle = {
    backgroundImage: imageUrl ? "url(".concat(imageUrl, ")") : 'none',
    backgroundColor: assignedColor,
    color: textColor
  };
  return React.createElement("div", _extends({
    className: classes,
    style: avatarStyle,
    "aria-label": name,
    title: name
  }, rest), optionalInitial);
}; // TODO: Migrate to a service

function checkValidColor(props, propName, componentName) {
  var validHex = /(^#[0-9A-F]{6}$)|(^#[0-9A-F]{3}$)/i.test(props.color);

  if (props.color && !validHex) {
    throw new Error("".concat(componentName, " needs to pass a valid color. This can either be a three ") + "or six character hex value");
  }
}

function checkValidInitials(props, propName, componentName) {
  // Must be the number "1" or "2"
  if (props.initialsLength && props.initialsLength > 2) {
    throw new Error("".concat(componentName, " only accepts 1 or 2 for the initials as a number"));
  } // Must be a string of 1 or 2 characters


  if (props.initials && props.initials.length > 2) {
    throw new Error("".concat(componentName, " only accepts a max of 2 characters for the initials as a string"));
  }
}

EuiAvatar.propTypes = {
  className: PropTypes.string,
  imageUrl: PropTypes.string,
  size: PropTypes.oneOf(SIZES),

  /**
   * Full name of avatar for title attribute and calculating initial if not provided
   */
  name: PropTypes.string.isRequired,

  /**
   * Accepts hex value `#FFFFFF`, `#000` otherwise a viz palette color will be assigned
   */
  color: checkValidColor,

  /**
   * Specify how many characters to show (max 2 allowed).
   * By default, will show based on number of words.
   */
  initialsLength: checkValidInitials,

  /**
   * Custom initials (max 2 characters).
   * By default will take the first character (of each word).
   */
  initials: checkValidInitials,

  /**
   * The type of avatar to this is displaying
   */
  type: PropTypes.oneOf(TYPES)
};
EuiAvatar.defaultProps = {
  size: 'm',
  type: 'user'
};
EuiAvatar.__docgenInfo = {
  "description": "",
  "methods": [],
  "displayName": "EuiAvatar",
  "props": {
    "size": {
      "defaultValue": {
        "value": "'m'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"none\"",
          "computed": false
        }, {
          "value": "\"s\"",
          "computed": false
        }, {
          "value": "\"m\"",
          "computed": false
        }, {
          "value": "\"l\"",
          "computed": false
        }, {
          "value": "\"xl\"",
          "computed": false
        }]
      },
      "required": false,
      "description": ""
    },
    "type": {
      "defaultValue": {
        "value": "'user'",
        "computed": false
      },
      "type": {
        "name": "enum",
        "value": [{
          "value": "\"space\"",
          "computed": false
        }, {
          "value": "\"user\"",
          "computed": false
        }]
      },
      "required": false,
      "description": "The type of avatar to this is displaying"
    },
    "className": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "imageUrl": {
      "type": {
        "name": "string"
      },
      "required": false,
      "description": ""
    },
    "name": {
      "type": {
        "name": "string"
      },
      "required": true,
      "description": "Full name of avatar for title attribute and calculating initial if not provided"
    },
    "color": {
      "type": {
        "name": "custom",
        "raw": "checkValidColor"
      },
      "required": false,
      "description": "Accepts hex value `#FFFFFF`, `#000` otherwise a viz palette color will be assigned"
    },
    "initialsLength": {
      "type": {
        "name": "custom",
        "raw": "checkValidInitials"
      },
      "required": false,
      "description": "Specify how many characters to show (max 2 allowed).\nBy default, will show based on number of words."
    },
    "initials": {
      "type": {
        "name": "custom",
        "raw": "checkValidInitials"
      },
      "required": false,
      "description": "Custom initials (max 2 characters).\nBy default will take the first character (of each word)."
    }
  }
};
