// x-pack/plugins/beats_management/public/pages/tag/index.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
require("brace/mode/yaml");
require("brace/theme/github");
const eui_1 = require("@elastic/eui");
const euiVars = tslib_1.__importStar(require("@elastic/eui/dist/eui_theme_k6_light.json"));
const lodash_1 = require("lodash");
const react_1 = tslib_1.__importDefault(require("react"));
const constants_1 = require("x-pack/plugins/beats_management/common/constants");
const primary_1 = require("../../components/layouts/primary");
const tag_1 = require("../../components/tag");
const with_url_state_1 = require("../../containers/with_url_state");
class TagPageComponent extends react_1.default.PureComponent {
    constructor(props) {
        super(props);
        this.mode = 'create';
        this.loadTag = async () => {
            const tags = await this.props.libs.tags.getTagsWithIds([this.props.match.params.tagid]);
            if (tags.length === 0) {
                // TODO do something to error
            }
            this.setState({
                tag: tags[0],
            });
        };
        this.loadAttachedBeats = async () => {
            const beats = await this.props.libs.beats.getBeatsWithTag(this.props.match.params.tagid);
            this.setState({
                attachedBeats: beats,
            });
        };
        this.saveTag = async () => {
            await this.props.libs.tags.upsertTag(this.state.tag);
            this.props.goTo(`/overview/tags`);
        };
        this.getNumExclusiveConfigurationBlocks = () => this.state.tag.configuration_blocks
            .map(({ type }) => constants_1.UNIQUENESS_ENFORCING_TYPES.some(uniqueType => uniqueType === type))
            .reduce((acc, cur) => (cur ? acc + 1 : acc), 0);
        const randomColor = lodash_1.sample(Object.keys(euiVars)
            .filter(key => key.startsWith('euiColorVis'))
            .map(key => euiVars[key]));
        this.state = {
            showFlyout: false,
            attachedBeats: null,
            tag: {
                id: props.match.params.action === 'create' ? '' : props.match.params.tagid,
                color: this.rgb2hex(randomColor),
                configuration_blocks: [],
                last_updated: new Date(),
            },
        };
        if (props.match.params.action !== 'create') {
            this.mode = 'edit';
            this.loadTag();
            this.loadAttachedBeats();
        }
    }
    render() {
        return (react_1.default.createElement(primary_1.PrimaryLayout, { title: this.mode === 'create' ? 'Create Tag' : `Update Tag: ${this.state.tag.id}` },
            react_1.default.createElement("div", null,
                react_1.default.createElement(tag_1.TagEdit, { tag: this.state.tag, mode: this.mode, onDetachBeat: async (beatIds) => {
                        await this.props.libs.beats.removeTagsFromBeats(beatIds.map(id => {
                            return { beatId: id, tag: this.state.tag.id };
                        }));
                        await this.loadAttachedBeats();
                    }, onTagChange: (field, value) => this.setState(oldState => ({
                        tag: { ...oldState.tag, [field]: value },
                    })), attachedBeats: this.state.attachedBeats }),
                react_1.default.createElement(eui_1.EuiSpacer, { size: "m" }),
                react_1.default.createElement(eui_1.EuiFlexGroup, null,
                    react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiButton, { fill: true, disabled: this.state.tag.id.search(/^[a-zA-Z0-9-]+$/) === -1 ||
                                this.state.tag.id === '' ||
                                this.getNumExclusiveConfigurationBlocks() > 1 // || this.state.tag.configuration_blocks.length === 0
                            , onClick: this.saveTag }, "Save")),
                    react_1.default.createElement(eui_1.EuiFlexItem, { grow: false },
                        react_1.default.createElement(eui_1.EuiButtonEmpty, { onClick: () => this.props.goTo('/overview/tags') }, "Cancel"))))));
    }
    rgb2hex(rgb) {
        const matchedrgb = rgb.match(/^rgba?[\s+]?\([\s+]?(\d+)[\s+]?,[\s+]?(\d+)[\s+]?,[\s+]?(\d+)[\s+]?/i);
        return matchedrgb && matchedrgb.length === 4
            ? '#' +
                ('0' + parseInt(matchedrgb[1], 10).toString(16)).slice(-2) +
                ('0' + parseInt(matchedrgb[2], 10).toString(16)).slice(-2) +
                ('0' + parseInt(matchedrgb[3], 10).toString(16)).slice(-2)
            : '';
    }
}
exports.TagPageComponent = TagPageComponent;
exports.TagPage = with_url_state_1.withUrlState(TagPageComponent);
