// x-pack/plugins/apm/server/routes/services.js
"use strict";
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const boom_1 = tslib_1.__importDefault(require("boom"));
const input_validation_1 = require("../lib/helpers/input_validation");
const setup_request_1 = require("../lib/helpers/setup_request");
const get_service_1 = require("../lib/services/get_service");
const get_services_1 = require("../lib/services/get_services");
const ROOT = '/api/apm/services';
const pre = [{ method: setup_request_1.setupRequest, assign: 'setup' }];
const defaultErrorHandler = (reply) => (err) => {
    // tslint:disable-next-line
    console.error(err.stack);
    // @ts-ignore
    reply(boom_1.default.wrap(err, 400));
};
function initServicesApi(server) {
    server.route({
        method: 'GET',
        path: ROOT,
        config: {
            pre,
            validate: {
                query: input_validation_1.withDefaultValidators()
            }
        },
        handler: (req, reply) => {
            const { setup } = req.pre;
            return get_services_1.getServices(setup)
                .then(reply)
                .catch(defaultErrorHandler(reply));
        }
    });
    server.route({
        method: 'GET',
        path: `${ROOT}/{serviceName}`,
        config: {
            pre,
            validate: {
                query: input_validation_1.withDefaultValidators()
            }
        },
        handler: (req, reply) => {
            const { setup } = req.pre;
            const { serviceName } = req.params;
            return get_service_1.getService(serviceName, setup)
                .then(reply)
                .catch(defaultErrorHandler(reply));
        }
    });
}
exports.initServicesApi = initServicesApi;
